packages/ra-core/src/reducer/admin/resource/data.js

```diff
@@ -69,6 +71,42 @@ export default (previousState = initialState, { type, payload, meta }) => {
             .map(record => ({ ...record, ...payload.data }));
         return addRecords(updatedRecords, previousState);
     }
+    if (type === CRUD_DELETE_OPTIMISTIC) {
+        const newState = Object.keys(previousState).reduce((acc, id) => {
+            if (id === payload.id) {
+                return acc;
+            }
+
+            return {
+                ...acc,
+                [id]: previousState[id],
+            };
+        }, {});
+
+        Object.defineProperty(newState, 'fetchedAt', {
+            value: previousState.fetchedAt,
+        });
+
+        return newState;
+    }
+    if (type === CRUD_DELETE_MANY_OPTIMISTIC) {
+        const newState = Object.keys(previousState).reduce((acc, id) => {
```

Comment:
For clarity, I'd rewrite this one with a `filter` and a simpler `reduce`, as follows:

```js
const newState = Object.entries(previousState)
   .filter(([key]) => !payload.ids.includes(key))
   .reduce((obj, [key, val]) => ({ ...obj, [key]: val }), {});
```